### ISSUE TITLE
Inconsistent 2D vs 3D interpolation behavior for mode 'reflect'

### ISSUE THREAD

**Issue by grlee77** [scikit-image/scikit-image#1585]
Currently, interpolation gives different results than one would expect in the 3D case for mode reflect.  Likely this is a problem in `ndi.map_coordinates` rather than scikit-image itself, but a potential workaround I propose below is to call map_coordinates with mode 'mirror' instead of reflect. 

Here is the output of `test_resize2d` from skimage/transforms/tests/test_warps.py if I change the mode from 'constant' to 'reflect'.  This tests still passes and looks as expected:
![reflect_2d](https://cloud.githubusercontent.com/assets/6528957/8573443/9395e116-255f-11e5-9b38-6dd2b1a3aacd.png)

However, for `test_resize3d_resize`, the output is not as expected, and the test fails:
![reflect_3d](https://cloud.githubusercontent.com/assets/6528957/8573402/502df7f6-255f-11e5-98ca-9523377857d5.png)

If I change the mode passed to ndi.map_coordinates to 'mirror', rather than 'reflect', the expected result is restored.  
![mirror_3d](https://cloud.githubusercontent.com/assets/6528957/8573467/b44e5744-255f-11e5-8807-c65584dd5710.png)

The code producing the above figures is as follows:

``` python
import numpy as np
from skimage.transform import resize
from matplotlib import pyplot as plt


""" 2D reflect test """

# copied from test_resize2d()
x = np.zeros((5, 5), dtype=np.double)
x[1, 1] = 1
resized = resize(x, (10, 10), order=0, mode='reflect')
ref = np.zeros((10, 10))
ref[2:4, 2:4] = 1

fig, axes = plt.subplots(1, 3)
axes[0].imshow(x, interpolation='nearest', cmap=plt.cm.gray)
axes[0].set_axis_off()
axes[0].set_title("original")
axes[1].imshow(resized, interpolation='nearest', cmap=plt.cm.gray)
axes[1].set_axis_off()
axes[1].set_title("resized")
axes[2].imshow(ref, interpolation='nearest', cmap=plt.cm.gray)
axes[2].set_axis_off()
axes[2].set_title("expected result")


""" 3D reflect test """

# copied from test_resize3d_resize()
x = np.zeros((5, 5, 3), dtype=np.double)
x[1, 1, :] = 1
resized = resize(x, (10, 10, 1), order=0, mode='reflect')
ref = np.zeros((10, 10, 1))
ref[2:4, 2:4] = 1

fig, axes = plt.subplots(1, 3)
axes[0].imshow(x[:, :, 1], interpolation='nearest', cmap=plt.cm.gray)
axes[0].set_axis_off()
axes[0].set_title("original[:, :, 1]")
axes[1].imshow(resized[:, :, 0], interpolation='nearest', cmap=plt.cm.gray)
axes[1].set_axis_off()
axes[1].set_title("resized")
axes[2].imshow(ref[:, :, 0], interpolation='nearest', cmap=plt.cm.gray)
axes[2].set_axis_off()
axes[2].set_title("expected result")


""" 3D mirror test """

# copied from test_resize3d_resize()
x = np.zeros((5, 5, 3), dtype=np.double)
x[1, 1, :] = 1
resized = resize(x, (10, 10, 1), order=0, mode='mirror')
ref = np.zeros((10, 10, 1))
ref[2:4, 2:4] = 1

fig, axes = plt.subplots(1, 3)
axes[0].imshow(x[:, :, 1], interpolation='nearest', cmap=plt.cm.gray)
axes[0].set_axis_off()
axes[0].set_title("original[:, :, 1]")
axes[1].imshow(resized[:, :, 0], interpolation='nearest', cmap=plt.cm.gray)
axes[1].set_axis_off()
axes[1].set_title("resized")
axes[2].imshow(ref[:, :, 0], interpolation='nearest', cmap=plt.cm.gray)
axes[2].set_axis_off()
axes[2].set_title("expected result")

```

**Comment by emmanuelle** [comment]
@grlee77  Was this issue solved by #1583? Can we close it?

**Comment by grlee77** [comment]
Not quite yet.  Now that #1583 has been merged I need to submit a small PR to change the default mode from `constant` to `symmetric` as discussed in #1514.

**Comment by grlee77** [comment]
I think due to some bugs in `ndimage.map_coordinates`, it will not be possible to resolve this for all edge cases.  The cleanest solution would be to update the skimage interpolation code to work for 3D as well, but I don't have the free time to work on that in the immediate future.  

From current master, edge mode 'symmetric' results in disagreement for 2D and 3D images in the example above.  For the 'symmetric' case (equivalent of 'reflect' in ndimage) I get the following questionable output from `ndimage` in a simple 1D example with equivalent x-coordinate range as the 2D example above:

```python
import scipy.ndimage as ndi
import numpy as np
x = np.arange(5.)
coords =  [np.arange(-0.25, 4.5, 0.25), ]    # 0.5 * (x - 0.5) + 0.5
ndi.map_coordinates(x, coords, order=0, mode='reflect')
```
outputs
```python
array([1, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4])
```
I think that first value should be a 0

If I extend the coordinate range further, this becomes more apparent
```python
In[57]:  ndi.map_coordinates(np.arange(5.), [np.arange(-2.25, 4.5, 0.25)], order=0, mode='reflect')
Out[57]: 
array([ 1.,  1.,  1.,  1.,  0.,  0.,  0.,  0.,  1.,  0.,  0.,  1.,  1.,
        1.,  1.,  2.,  2.,  2.,  2.,  3.,  3.,  3.,  3.,  4.,  4.,  4.,  4.])
```

**Comment by stefanv** [comment]
Wouldn't fixing this directly in ``scipy.ndimage`` be the easiest
approach?  I'd be happy to help review and merge such a PR.

**Comment by sciunto** [comment]
@grlee77 Would you like to open a bug report to scipy?

**Comment by grlee77** [comment]
@sciunto.  There is a related issue open at scipy/scipy#2640 
I did spend a few hours looking at it some time ago, but did not come up with a fix that didn't also break some other existing tests.  I doubt I will have the time to revisit it in the near future, but would definitely like to sort it out at some point.

**Comment by sciunto** [comment]
Thanks for the link.

**Comment by rfezzani** [comment]
I can't reproduce this one. Closing! ;-)

**Comment by stefanv** [comment]
Checking in with @grlee77 as to whether he considers this resolved.

**Comment by grlee77** [comment]
Actually, this particular inconsistency between 2D and 3D still exists, it just no longer occurs for the default `"reflect"` mode due to the renaming for consistency with `numpy.pad` that took place in #1583.

An updated code to reproduce the issue is given below by swapping out the name 'reflect' for 'symmetric'

```Python
# 2D symmetric test

x = np.zeros((5, 5), dtype=np.double)
x[1, 1] = 1
resized = resize(x, (10, 10), order=0, mode='symmetric')
ref = np.zeros((10, 10))
ref[2:4, 2:4] = 1

fig, axes = plt.subplots(1, 3)
axes[0].imshow(x, interpolation='nearest', cmap=plt.cm.gray)
axes[0].set_axis_off()
axes[0].set_title("original")
axes[1].imshow(resized, interpolation='nearest', cmap=plt.cm.gray)
axes[1].set_axis_off()
axes[1].set_title("resized")
axes[2].imshow(ref, interpolation='nearest', cmap=plt.cm.gray)
axes[2].set_axis_off()
axes[2].set_title("expected result")


""" 3D symmetric test """

# copied from test_resize3d_resize()
x = np.zeros((5, 5, 3), dtype=np.double)
x[1, 1, :] = 1
resized = resize(x, (10, 10, 1), order=0, mode='symmetric')
ref = np.zeros((10, 10, 1))
ref[2:4, 2:4] = 1

fig, axes = plt.subplots(1, 3)
axes[0].imshow(x[:, :, 1], interpolation='nearest', cmap=plt.cm.gray)
axes[0].set_axis_off()
axes[0].set_title("original[:, :, 1]")
axes[1].imshow(resized[:, :, 0], interpolation='nearest', cmap=plt.cm.gray)
axes[1].set_axis_off()
axes[1].set_title("resized")
axes[2].imshow(ref[:, :, 0], interpolation='nearest', cmap=plt.cm.gray)
axes[2].set_axis_off()
axes[2].set_title("expected result")
```

There is still the related upstream issue that `scipy.ndimage`'s wrap differs from `np.pad` or OpenCV as illustrated in this comment: https://github.com/scikit-image/scikit-image/pull/1583#issuecomment-127772723

**Comment by rfezzani** [comment]
OK, sorry, I see the problem now, it comes from the mismatch between scipy and skimage  behavior on 'symmetric' mode. Isn't the `wrap` mode also affected according to your https://github.com/scipy/scipy/issues/2640#issuecomment-120079980?

**Comment by grlee77** [comment]
I have confirmed that this issue is closed by scipy/scipy#12767 which will be included in the upcoming 1.6.0 release in a couple of weeks.

This one has been a long time coming (5.5 years!)

**Comment by jni** [comment]
So happy about this one @grlee77! ❤️
